exec/cnex/number.c

```diff
@@ -17,49 +17,125 @@
+    const int trailing_zeros = (int)slen - last_significant_digit;
+    exponent += trailing_zeros;
+    char sbuf[50] = { 0 };
+    char r[50] = { 0 };
+    // Copy the mantissa out to the last significant digit.
+    memcpy(sbuf, v, last_significant_digit);
+    // Guarantee the string is null terminated, despite the buffer starting off NULL.
+    sbuf[last_significant_digit] = '\0';
+
+    if (exponent != 0) {
+        if (exponent > 0 && last_significant_digit + exponent <= PRECISION) {
+            // We have an exponent value we need to deal with, but the length combined with the exponent
+            // is less than our desired max precision, so we move the sbuf value into (r)esult, the length
+            // of the significant mantissa, adding zeros as necessary.
+            memcpy(r, sbuf, last_significant_digit);
+            for (int i = last_significant_digit; i < (last_significant_digit + exponent); i++) {
+                r[i] = '0';
+            }
+        } else if (exponent < 0 && -exponent < slen) {
+            // Our exponent is negative, and is less than the length of the mantissa, so we're
+            // going to insert the decimal into slen+exponent location.
+            memcpy(r, sbuf, slen+exponent);
+            r[slen+exponent] = '.';
+            memcpy(&r[slen+exponent]+1, &sbuf[slen+exponent], slen - (slen+exponent));
+        } else if (exponent < 0 && -exponent == slen) {
+            // Our exponent is negative, and is the length of the mantissa, so we're going to start
+            // the number with "0.", and then add the rest of the mantissa on the right side of the decimal.
+            memcpy(r, "0.", 2);
+            memcpy(&r[2], sbuf, slen);
+        } else if (exponent < 0 && slen - exponent <= PRECISION) {
+            // Our exponent is negative, and the mantissa length, minus the exponent less or equal to our max PRECISION
+            // then we need to insert zero's to the right of the decimal.
+            int count = 2;
+            // Start out by setting [r]esult to "0."
+            memcpy(r, "0.", count);
+            // Now, we copy the necessary number of zeros after the decimal point...
+            for (int i = 0; i < -exponent-slen;i++, count++) {
+                r[count] = '0';
+            }
+            // Then copy the remainder of the mantissa to our [r]esult.
+            memcpy(&r[count], sbuf, slen);
+            // Ensure that [r]esult is null terminated.  I'm not 100% convinced this is necessary.
+            r[count+slen] = '\0';
+        } else {
+            // Our exponent is positive, therefore we need to add the decimal point on the right of the number.
+            exponent += last_significant_digit - 1;
+            if (last_significant_digit >= 3) {
+                // Our mantissa is at least 3 digits long, so copy the first digit to [r]esult
+                memcpy(r, sbuf, 1);
```

Comment:
It's probably clearer to do a single assignment statement here rather than a `memcpy` of just one byte.

Reply:
Fixed.